usb/core.py

```diff
@@ -770,13 +771,32 @@ def __init__(self, dev, backend):
             self.speed = None
 
     @property
+    def langids(self):
+        """ Return the USB device's supported language ID codes.
+
+        These are 16-bit codes familiar to Windows developers, where for
+        example instead of en-US you say 0x0409. USB_LANGIDS.pdf on the usb.org
+        developer site for more info. String requests using a LANGID not
+        in this array should not be sent to the device.
+
+        This property will cause some USB traffic the first time it is accessed
+        and cache the resulting value for future use.
+        """
+        if self._langids is None:
+            try:
+                self._langids = util.get_langids(self)
+            except USBError:
+                self._langids = array.array('H')
```

Comment:
Sorry ... I didn't squash these 4 commits because I thought the commit messages in order would help make the intention clear. This is from the fourth commit, handling the fact that some devices with no LANGIDs don't even return an empty LANGIDs array, they return _nothing_ and a backend pipe error results. Client code should treat the case the same as no LANGIDs supported, best achieved by returning the same thing (empty array) to client code in either case.